### PR TITLE
CommitlogArchiver only has granularity to seconds for restore_point_in_time

### PR DESCRIPTION
**Pull request by Maxwell-Guo**
```
CommitlogArchiver only has granularity to seconds for restore_point_in_time for [CASSANDRA-19448](https://issues.apache.org/jira/browse/CASSANDRA-19448)

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: MaxwellGuo
Co-authored-by:Tiago L. Alves
```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: conf/commitlog_archiving.properties [hunk 1]
```diff
@@ -37,7 +37,11 @@ restore_command=
 restore_directories=
 
 # Restore mutations created up to and including this timestamp in GMT.
-# Format: yyyy:MM:dd HH:mm:ss (2012:04:31 20:43:12)
+# There are only three different formats to express three time precisions:
+# Seconds, Milliseconds, and Microseconds.
+# Seconds format: yyyy:MM:dd HH:mm:ss (2012:04:31 20:43:12)
+# Milliseconds format: yyyy:MM:dd HH:mm:ss[.[SSS]] (2012:04:31 20:43:12.633)
+# Microseconds format: yyyy:MM:dd HH:mm:ss[.[SSSSSS]] (2012:04:31 20:43:12.633222)
```
Comment (tiagomlalves) [inline conf/commitlog_archiving.properties]: I would propose removing the brackets which indicate optional/groups, and just specify:
```
# Milliseconds format: yyyy:MM:dd HH:mm:ss.SSS (2012:04:31 20:43:12.633)
# Microseconds format: yyyy:MM:dd HH:mm:ss.SSSSSS (2012:04:31 20:43:12.633222)
```

File: src/java/org/apache/cassandra/db/commitlog/CommitLogArchiver.java [hunk 2]
```diff
@@ -103,62 +108,76 @@ public static CommitLogArchiver construct()
...
-
-                String snapshotPosition = commitlog_commands.getProperty("snapshot_commitlog_position");
-                CommitLogPosition snapshotCommitLogPosition;
-                try
-                {
+            }
+        }
+        String targetTime = commitlogCommands.getProperty("restore_point_in_time");
+        //todo remove this as this is not used
+        TimeUnit precision = TimeUnit.valueOf(commitlogCommands.getProperty("precision", "MICROSECONDS"));
```
Comment (tiagomlalves) [inline src/java/org/apache/cassandra/db/commitlog/CommitLogArchiver.java]: agree :)

File: src/java/org/apache/cassandra/db/commitlog/CommitLogArchiver.java [hunk 3]
```diff
@@ -103,62 +108,76 @@ public static CommitLogArchiver construct()
...
+            else
+            {
+                // get restorePointInTime  microlevel by default as c* use this level ts.
+                restorePointInTime = getMicroSeconds(targetTime);
+            }
+        }
+        catch (DateTimeParseException | ConfigurationException e)
+        {
+            throw new RuntimeException("Unable to parse restore target time", e);
+        }
```
Comment (tiagomlalves) [inline src/java/org/apache/cassandra/db/commitlog/CommitLogArchiver.java]: Where does `ConfigurationException` comes from?

Here, I would rewrite this as:
```
        long restorePointInTime = Long.MAX_VALUE;
        try
        {
            if (!Strings.isNullOrEmpty(targetTime))
            {
                // get restorePointInTime  microlevel by default as c* use this level ts.
                restorePointInTime = getMicroSeconds(targetTime);
            }
        }   
```
or just (which is also used below):
```
        long restorePointInTime;
        try
        {
            restorePointInTime = Strings.isNullOrEmpty(targetTime)
                ? Long.MAX_VALUE
                :  getMicroSeconds(targetTime);
        }
```

to reduce the nesting.

File: src/java/org/apache/cassandra/db/commitlog/CommitLogMBean.java [hunk 4]
```diff
@@ -41,7 +41,11 @@ public interface CommitLogMBean
 
     /**
      * Restore mutations created up to and including this timestamp in GMT
-     * Format: yyyy:MM:dd HH:mm:ss (2012:04:31 20:43:12)
+     * There are only three different formats to express three time precisions:
+     * Seconds, Milliseconds, and Microseconds.
+     * Seconds format: yyyy:MM:dd HH:mm:ss (2012:04:31 20:43:12)
+     * Milliseconds format: yyyy:MM:dd HH:mm:ss[.[SSS]] (2012:04:31 20:43:12.633)
+     * Microseconds format: yyyy:MM:dd HH:mm:ss[.[SSSSSS]] (2012:04:31 20:43:12.633222)
```
Comment (tiagomlalves) [inline src/java/org/apache/cassandra/db/commitlog/CommitLogMBean.java]: same comment as above

File: test/unit/org/apache/cassandra/cql3/validation/operations/DropRecreateAndRestoreTest.java [hunk 5]
```diff
@@ -71,13 +71,13 @@ public void testCreateWithIdRestore() throws Throwable
             FileUtils.renameWithConfirm(new File(logPath, segment + ".save"), new File(logPath, segment));
         try
         {
-            // Restore to point in time.
-            CommitLog.instance.archiver.restorePointInTime = time;
+            // Restore to point in time, the rpi time is micro level
```
Comment (tiagomlalves) [inline test/unit/org/apache/cassandra/cql3/validation/operations/DropRecreateAndRestoreTest.java]: what about?
```
// Restore to point in time (microseconds granularity).
```

File: test/unit/org/apache/cassandra/db/RecoveryManagerTest.java [hunk 6]
```diff
@@ -250,28 +249,84 @@ public void testRecoverCounter() throws IOException
     public void testRecoverPIT() throws Exception
     {
         CommitLog.instance.resetUnsafe(true);
+        long rpiTimeInit = CommitLog.instance.archiver.restorePointInTimeInMicros;
```
Comment (tiagomlalves) [inline test/unit/org/apache/cassandra/db/RecoveryManagerTest.java]: Whenever I see `rpi` my brain parses it as Raspberry Pi 🤓 sorry!
What calling it `originalPIT` ? Later we'll be doing `.setRestorePointInTime(originalPIT);`

**Review by tiagomlalves — APPROVED** [review]
Looks great! Thanks

File: src/java/org/apache/cassandra/db/commitlog/CommitLogArchiver.java (hunk 2 above)
Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/db/commitlog/CommitLogArchiver.java]: @tiagomlalves Let me confirm with you, does your "agree" mean agree with "todo remove this as this is not used "?

**Comment by kenjords** [comment]
@Maxwell-Guo Thank you  folk for looking into this issue, when do we think that this patch will be merged in?

**Comment by Maxwell-Guo** [comment]
hi @kenjords ，I have been very busy in recent months, so I have not pushed forward with this work. I believe that I will have time to continue this work next week. In addition, this work should be nearing completion.

**Pull request closed without merging**